book/src/rust/rust_style_guide.md

```diff
@@ -0,0 +1,660 @@
+6. For a given dependency, use the following structure with optional and features keys as needed:
+
+```toml
+crate-name = { version = "...", optional = true, default-features = false, features = ["..."] }
+If the crate is a path or git dependency, replace those keys with the version key and add a tag, branch, or rev as needed following the git key.
+```
+7. When adding a feature, add a doc string in the title case and a new line between each feature.
+
+#### Feature Selection
+
+When using features, be sure to attach a doc_cfg feature declaration as well unless the code is not exported to pub.
+
+```rust
+#[cfg(feature = "...")]
+#[cfg_attr(doc_cfg, doc(cfg(feature = "...")))]
+pub mod feature_gated_public_module;
+```
+
+#### Imports and Exports
+
+Imports (use) and exports (mod) should be ordered as follows:
+
+1. External Crate Declarations
+
+2. Private Imports
+
+3. Private Imports with Features
+
+4. Private Exports
+
+5. Private Exports with Features
+
+6. Public Exports
+
+7. Public Exports with Features
+
+8. Reexports
+
+9. Reexports with Features
+
+Here's an example set of declarations:
+
+```rust
+extern crate crate_name;
+
+use module::submodule::entry;
+
+#[cfg(feature = "...")]
+use module::feature_gated_submodule;
+
+mod another_module;
+mod module;
+mod the_third_module;
+
+#[cfg(feature = "...")]
+mod feature_gated_module;
+
+pub mod public_module;
+
+#[cfg(feature = "...")]
+#[cfg_attr(doc_cfg, doc(cfg(feature = "...")))]
+pub mod feature_gated_public_module;
+
+pub use reexported_objects;
+
+#[cfg(feature = "...")]
+#[cfg_attr(doc_cfg, doc(cfg(feature = "...")))]
+pub use feature_gated_reexported_objects;
+```
+
+Ensure that there are new lines between each category. Be sure that if some imports or exports are feature-gated, that they are sorted by feature alphabetically. If there is a feature gated import that requires importing multiple objects, use the following pattern:
+
+```rust
+#[cfg(feature = "...")]
+use {
+    thing1, thing2, thing3, thing4,
+};
+```
+
+**NOTE**: All imports should occur at the top of any module, and a newline should be added between the last import and the first declared object.
+
+#### Let binding
+
+
+##### Always separately bind [RAII](https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization) guards
+
+
+>RAII stands for "Resource Acquisition is Initialisation". The essence of the pattern is that resource initialisation is done in the constructor of an object and finalisation in the destructor. This pattern is extended in Rust by using a RAII object as a guard of some resource and relying on the type system to ensure that access is always mediated by the guard object.
+
+Prefer
+
+```rust
+fn use_mutex(m: sync::mutex::Mutex<int>) {
+    let guard = m.lock();
+    do_work(guard);
+    drop(guard); // unlock the lock
+    // do other work
+}
+```
+over
+
+```rust
+fn use_mutex(m: sync::mutex::Mutex<int>) {
+    do_work(m.lock());
+    // do other work
+}
+```
```

Comment:
I don't think this really adds much. This made more sense pre-NLL, but IMO this should be left to the developer's judgement

Reply:
Why is the first one preferred? I usually find the second one less error-prone.

@cameron1024 What is "pre-NLL"?

Reply:
I think the issue could be that in the second the drop is implicit, it could happen at the end of the use_mutex function,  not necessarily immediately after the lock is no longer needed.  Is it defined by the language when the drop will occur?  Dropping the guard explicitly ensures that the mutex is released as early as it can be.  Its also still safe because if the code was:

```rust
fn use_mutex(m: sync::mutex::Mutex<int>) -> Option<usize> {
    let guard = m.lock();
    do_work(guard)?;
    drop(guard); // unlock the lock
    // do other "Long running" work
    let result = ...

    Some(result)
}
```

The lock would get dropped automatically when do_work returned None, because it would immediately go out of scope.